### PR TITLE
wrapInstance: Search inheritance hierarchy for a suitable class

### PR DESCRIPTION
**Pull request by mitchellsimmons**
This is in reference to #348.

Adds support to QtCompat.wrapInstance for classes which have a deeper inheritance hierarchy.

## Example

Calling QtCompat.wrapInstance with a pointer to an instance of class B will now return a QDialog instead of a QObject.

```python
class A(QtWidgets.QDialog):
    pass
    
class B(A):
    pass
```

### CODE REVIEW

**Comment by mottosso** [comment]
I think this looks great, thanks @mitchellsimmons.

From the test suite, it looks like nothing existing broke. That's great. The remaining bit is converting your example into a test, [here](https://github.com/mottosso/Qt.py/blob/cfd1b1d3155485d7b855b9578fc847cf7f5a5153/tests.py#L899). Something that exercises this feature and covers as many corner cases as you can think of.

Up for it?

**Comment by mitchellsimmons** [comment]
Thanks @mottosso.

I ran into an issue when trying to run the docker image.
The entrypoint.sh file appears to be using CRLF line endings.
I had to convert to LF to get the tests to run.

Once I've made my changes what would be the best way for me to post them here?

**Comment by mitchellsimmons** [comment]
Sorry you can ignore the line endings issue, this was my fault. I'm using windows so I needed to update `git config --global core.autocrlf` to `false` before cloning the repository.  Maybe it could be useful to add a `.gitattributes` file to force LF line endings for the entrypoint.sh file.

**Comment by mottosso** [comment]
> Once I've made my changes what would be the best way for me to post them here?

Whenever you push to your own repository, this PR will automatically update and re-run the tests with your latest changes. So go right ahead and push. And don't worry about pushing too often and creating too many commits. If it becomes a problem, we can "squash" all of your commits into a single one for a simplified Git history. That way you also don't necessarily need to run the tests on your own machine/Docker, although that is probably the faster option since it can take a minute or two for GitHub to get started.

**Comment by mitchellsimmons** [comment]
Thanks for explaining this @mottosso.

Let me know what you think of the tests, I can make some changes if needed.

Regarding the minor change to `wrapInstance`  [here](https://github.com/mitchellsimmons/Qt.py/blob/9fdd65aec8efd9ef466e08f00e80d096c057e305/Qt.py#L782): after running my tests with sip I found that it always returns the most suitable class regardless of what the user provides so i decided it isn't necessary to run the search when sip is being used.

Regarding changes to testing [here](https://github.com/mitchellsimmons/Qt.py/blob/9fdd65aec8efd9ef466e08f00e80d096c057e305/tests.py#L308) and [here](https://github.com/mitchellsimmons/Qt.py/blob/9fdd65aec8efd9ef466e08f00e80d096c057e305/tests.py#L894): tests are not currently running for Python 3.4 because of the micro version.

File: tests.py [hunk 1]
```diff
@@ -920,8 +932,14 @@ def test_wrapInstance():
         finally:
             app.exit()
 
-    def test_implicit_wrapInstance():
-        """.wrapInstance doesn't need the `base` argument"""
+    def test_implicit_wrapInstance_for_base_types():
+        """Tests implicit `Foo` -> `Foo` cast of `Foo` pointer to `Foo` object,
```
Comment (mottosso) [inline tests.py:936]: This summary is a little long, and has a line-break.

The first line of the docstring is what shows up as [the name of the test](https://travis-ci.org/github/mottosso/Qt.py/jobs/724576499#L576) so we'd better keep it short and sweet. You can move the elaboration to the next line and beyond. Looks pretty good already perhaps, without the comma at the end there.

File: tests.py [hunk 2]
```diff
@@ -920,8 +932,14 @@ def test_wrapInstance():
...
+        produced by .wrapInstance when:
+
+        #. The `base` argument has a default value.
```
Comment (mottosso) [inline tests.py:939]: Are these bullet lists? Would rather make them `1.` and `2.` in that case to keep focus on the code rather than the formatting.

**Review by mottosso — CHANGES_REQUESTED** [review]
Looks good to me, great work. Other than a few style-issues, this is good to merge.

File: tests.py (hunk 2 above)
Comment (mitchellsimmons) [reply tests.py]: Yep it's just rst auto-enumeration, I've changed it to explicit.
If any of the elaboration is too much I'm happy to remove it.
Cheers

Comment (mottosso) [reply tests.py]: Ooo, learnt something new. Thanks.

**Review by mottosso — APPROVED** [review]

**Comment by mottosso** [comment]
Thanks @mitchellsimmons, it's a wrap!
